INT-4105: RedisLock: unlock local in `obtain()`

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4105
Fixes GH-1888

The lock in Redis can be expired in between `obtain()` calls.
So, even if we return a new lock instance, the old one must clear properly.
- Add `lock.unlock()` to the `obtain()` if the case of expiration in Redis.
- Add `warn` for the exception on the `lock.unlock()`. We don't care about error here and just proceed to a new instance.

**Chery-pick to 4.3.x**

artembilan committed:
INT-4105: RedisLock: unlock local in `obtain()`

JIRA: https://jira.spring.io/browse/INT-4105
Fixes GH-1888

The lock in Redis can be expired in between `obtain()` calls.
So, even if we return a new lock instance, the old one must clear properly.

* Add `lock.unlock()` to the `obtain()` if the case of expiration in Redis.
* Add `warn` for the exception on the `lock.unlock()`. We don't care about error here and just proceed to a new instance.

**Chery-pick to 4.3.x**

garyrussell commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java:
```diff
@@ -260,6 +260,14 @@ public Lock obtain(Object lockKey) {
+				}
+				catch (Exception e) {
+					if (logger.isWarnEnabled()) {
+						logger.warn("Lock was released due to expiration. Will be obtained a new one...", e);
```
logger.warn("Lock was released due to expiration. A new one will be obtained...", e);

If `useWeakReferences` is true, we'll find the same lock again below. Need to remove from that map as well under this condition.

Add `registry.setUseWeakReferences(true);` to your test and it fails.

artembilan committed:
Remove the lock reference from `weakThreadLocks` as well
Add `registry.setUseWeakReferences(true);` to test-case

rpresnakov commented on spring-integration-redis/src/test/java/org/springframework/integration/redis/util/RedisLockRegistryTests.java:
```diff
@@ -384,6 +385,18 @@ public void testExpireNoLockInStore() throws Exception {
 
 	@Test
 	@RedisAvailable
+	public void testExpireDuringSecondObtain() throws Exception {
```
Hi @artembilan ! As I understand correctly this test will pass even without your latest changes in the `RedisLockRegistry` class:  when you do `registry.obtain("foo")` in case the Redis key has expired it will return new `RedisLock` object anyway and `assertNotSame(foo, foo1)` will pass.

Here is another use case: 
  1. Thread A obtains LockFoo
  2. Thread A locks LockFoo
  3. LockFoo expired
  4. Another Thread B successfully obtains LockFoo (will never get into the `if (lockInStore == null || !lock.equals(lockInStore)) {` block)
  5. Thread B tries to lock LockFoo and is blocked because the ReentrantLock is still locked by Thread A

What should be the expected behaviour in your opinion? I think that Thread B should be able to obtain and lock the lock which has already expired in Redis.

artembilan replied:
Thank you for feedback, @rpresnakov !

I've just added one more assert there. See the current version of this test.

Regarding your use-case...

Sounds correct and as expected behavior in the same JVM with normal `Lock`.

Here we really can't know if the Redis value has been expired until we ask for that, like in case of `unlock()`.

That will work as you expect if you use different `RedisLockRegistry` instances.
Otherwise everything should look like you use `DefaultLockRegistry`.

We really can't unlock `java.util.concurrent.locks.ReentrantLock` from some other Thread, unless owner...

artembilan committed:
* Add assert for `lock.unlock()` in case of expiration

* Add `UUID.randomUUID()` for the registry key to avoid cross-talking during concurrent builds, e.g. on the CI server
https://build.spring.io/browse/INT-MASTER-352

This is actually a fix for the

JIRA: https://jira.spring.io/browse/INT-4083

garyrussell commented:
Merged as 450ac7f18d60714e3075b16db470ffd520da79dc and cherry-picked.
